tests/post/steps/test_authentication.py

```diff
@@ -0,0 +1,207 @@
+    pass
+
+
+@scenario('../features/authentication.feature',
+          'Login to Dex using incorrect email')
+def test_failed_login(host):
+    pass
+
+
+@scenario('../features/authentication.feature',
+          'Login to Dex using correct email and password')
+def test_login(host):
+    pass
+
+
+@pytest.fixture(scope='function')
+def context():
+    return {}
+
+# }}}
+
+# Given {{{
+
+# }}}
+
+
+# When {{{
+
+
+@when(parsers.parse(
+    "we perform a request on '{path}' with port '{port}' on control-plane IP"))
+def perform_request(host, context, path, port):
+    ip = _get_control_plane_ip(host)
+    try:
+        context['response'] = requests.get(
+            'https://{ip}:{port}{path}'.format(
+                ip=ip, port=port, path=path
+            ),
+            verify=False,
+        )
+    except requests.exceptions.ConnectionError as exc:
+        context['exception'] = exc
+
+
+@when(parsers.parse(
+    "we login to Dex as '{username}' using password '{password}'"))
+def dex_login(host, username, password, context):
+    ip = _get_control_plane_ip(host)
+    context['login_response'] = _dex_auth_request(
+        ip, username, password
+    )
+
+
+# }}}
+
+
+# Then {{{
+
+
+@then("we can reach the OIDC openID configuration")
+def reach_openid_config(host):
+    ip = _get_control_plane_ip(host)
+    try:
+        response = requests.get(
+            'https://{ip}:8443/oidc/.well-known/openid-configuration'.format(
+                ip=ip
+            ),
+            verify=False,
+        )
+    except requests.exceptions.ConnectionError as exc:
+        raise ConnectionError(exc)
+
+    assert response.status_code == 200
+    response_body = response.json()
+    # check for the existence of  keys[issuer, authorization_endpoint]
+    assert 'issuer' and 'authorization_endpoint' in response_body
+
+
+@then(parsers.parse(
+    "the server returns '{status_code}' with message '{status_message}'"))
+def server_returns(host, context, status_code, status_message):
+    response = context.get('response')
+    assert response is not None
+    assert response.status_code == int(status_code)
+    assert response.text.rstrip('\n') == status_message
+
+
+@then(parsers.parse("authentication fails with login error"))
+def failed_login(host, context):
+    auth_response = context.get('login_response')
+    assert auth_response.text is not None
+    assert auth_response.status_code == 200
+    #'Invalid Email Address and password' is found in auth_response.text
+    assert 'Invalid Email Address and password' in auth_response.text
+    assert auth_response.headers.get('location') is None
+
+
+@then(parsers.parse("the server returns '{status_code}' with an ID token"))
+def successful_login(host, context, status_code):
+    auth_response = context.get('login_response')
+    if auth_response.text is None:
+        assert False
+    assert auth_response.status_code == int(status_code)
+    assert auth_response.headers.get('location') is not None
```

Comment:
Can we check the value? It should be the callback URL, right?

Reply:
It is actually an `Approval URL`. It only occurs when login was successful and Redirect(callback) has happened. But since we do not have a proper in-browser or out-of-browser redirect, we can simply check that this field exist when login passes.